includes/amp-helper-functions.php

```diff
@@ -411,23 +411,60 @@ function amp_is_available() {
 	}
 
 	$warn = function () {
-		static $warned = false;
-		if ( $warned ) {
+		static $already_warned_sources = [];
+
+		// Gather the themes and plugins responsible for calling the function incorrectly.
+		$closest_source = [
+			'type' => '',
+			'name' => '',
+		];
+
+		foreach ( debug_backtrace( 0 ) as $call_stack ) { // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_debug_backtrace
+			if ( empty( $call_stack['file'] ) ) {
+				continue;
+			}
+
+			$source = AMP_Validation_Manager::get_file_source( $call_stack['file'] );
+			if ( empty( $source ) || 'core' === $source['type'] || ( 'plugin' === $source['type'] && 'amp' === $source['name'] ) ) {
+				continue;
+			}
+
+			$closest_source = $source;
+			break;
+		}
+
+		$closest_source_identifier = $closest_source['type'] . ':' . $closest_source['name'];
+		if ( in_array( $closest_source_identifier, $already_warned_sources, true ) ) {
 			return;
 		}
+
 		$message = sprintf(
-			/* translators: %1$s: amp_is_available(), %2$s: amp_is_request(), %3$s: is_amp_endpoint(), %4$s: the current action, %5$s: the wp action, %6$s: the WP_Query class, %7$s: the amp_skip_post() function */
-			__( '%1$s (or %2$s, formerly %3$s) was called too early and so it will not work properly. WordPress is currently doing the "%4$s" action. Calling this function before the "%5$s" action means it will not have access to %6$s and the queried object to determine if it is an AMP response, thus neither the "%7$s" filter nor the AMP enabled toggle will be considered.', 'amp' ),
-			'amp_is_available()',
-			'amp_is_request()',
-			'is_amp_endpoint()',
-			current_action(),
-			'wp',
-			'WP_Query',
-			'amp_skip_post()'
+			/* translators: 1: amp_is_available() function, 2: amp_is_request() function, 3: is_amp_endpoint() function */
+			__( '%1$s (or %2$s, formerly %3$s) was called too early and so it will not work properly.', 'amp' ),
+			'`amp_is_available()`',
+			'`amp_is_request()`',
+			'`is_amp_endpoint()`'
+		);
+
+		if ( ! empty( $closest_source['type'] ) && ! empty( $closest_source['name'] ) ) {
+			$message .= ' ' . sprintf(
+				/* translators: 1: type (theme, plugin, must-use plugin), 2: name */
+				__( 'It appears the %1$s with slug %2$s is responsible; please contact the author.', 'amp' ),
+				'`' . $closest_source['type'] . '`',
+				'`' . $closest_source['name'] . '`'
```

Comment:
Yes, I thought about that. But I decided to not go to that effort since this is an error message. The type will correspond to directory names that a WP site admin will already be familiar with, like `plugin`, `theme`, and `mu-plugin`. For that reason, I thought translating these would be overkill.

Reply:
In other words, WordPress doesn't localize its filesystem.

Reply:
A regular user creating content on their site might never have seen the WordPress filesystem. All they see is the translated menu entries and UI elements.

Reply:
Where would regular users see the `_doing_it_wrong()` notices? Wouldn't it be in the context of PHP errors? There isn't otherwise a UI for displaying this to regular users (yet).

I'm not against translating these. I was just cutting a corner which didn't seem super important.

Reply:
Right, they wouldn't see this for the `doing_it_wrong()` part. But I reused this for the `ErrorPage` which is shown if AMP processing failed. This can indeed be shown to regular users.

I can update the output in the `ErrorPage` independently.

Reply:
Oh yes, of course 👍 

One thing to keep keep in mind for translation is that there will probably need to be three translation strings:

* `It appears the plugin with slug %s is responsible.`
* `It appears the must-use plugin with slug %s is responsible.`
* `It appears the theme with slug %s is responsible.`

As I know you're intimately aware, the “the” will translate differently depending on what it is defining.

I think it's useful to go include the slug for sure, but also we could include the name. For development purposes, the slug is much more helpful since it makes it easy to know where to `cd`. For non-development purposes, the name is more helpful to identify something to deactivate.

Reply:
I adapted this for `ErrorPage` but then went ahead and modified the text in `is_amp_endpoint()` as well. Even though the translation was technically fine for `is_amp_endpoint()` in the expected developer context, it makes sense to have shared translations across these two.

Reply:
I didn't put the plugin/theme name into the string just yet and kept the slug for now. The reason is that the field in the `source` array that is used for storing the slug is called `name`. Adding the actual name in addition to the slug would therefore require a refactor that touches a lot of code, so this is best done in a separate PR. This one is messy enough as it is.